File: scripts/parse-figma-collection.mjs
/**
 * Parses Figma Collection frame metadata XML into structured token JSON.
 * Source: VDS Figma Collection (73102:139464)
 *
 * Usage: node scripts/parse-figma-collection.mjs [path-to-xml]
 */
import { readFileSync, writeFileSync, mkdirSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const defaultInput =
  process.argv[2] ||
  join(
    process.env.HOME,
    '.cursor/projects/Users-ankush-rustagi-Verkada-Code/agent-tools/9f913e5d-7a56-4475-8e89-b212a2181620.txt',
  )

const OUT_DIR = join(__dirname, '../src/tokens/generated')

function extractTextNames(cellXml) {
  return [...cellXml.matchAll(/<text[^>]* name="([^"]*)"/g)].map((m) => m[1])
}

function parseHex(cellXml) {
  const hexMatch = cellXml.match(/\(#([0-9A-Fa-f]{6,8})\)/)
  const stepMatch = cellXml.match(/<text[^>]* name="(\d+|transparent|scrim)"[^>]*\/>/)
  return {
    hex: hexMatch ? `#${hexMatch[1].toUpperCase()}` : null,
    step: stepMatch ? stepMatch[1] : null,
  }
}

function parseValueCell(cellXml) {
  const texts = extractTextNames(cellXml)
  const { hex, step } = parseHex(cellXml)
  // Non-color values: px, %, numbers
  const raw = texts.filter((t) => !t.startsWith('(') && t !== 'Rectangle').join(' ')
  return { hex, step, raw: raw || null }
}

function normalizeGroup(header) {
  return header.replace(/^--/, '').trim()
}

function buildTokenName(group, variable) {
  if (variable.startsWith('--')) return variable
  if (!group) return `--${variable}`
  const g = group.startsWith('--') ? group.slice(2) : group
  return `--${g}/${variable}`
}

function parseTable(tableXml, tableName) {
  const rows = []
  let currentGroup = ''

  const rowBlocks = [...tableXml.matchAll(/<frame id="[^"]*" name="(Header|Row)"[^>]*>([\s\S]*?)<\/frame>/g)]

  for (const [, kind, body] of rowBlocks) {
    if (kind === 'Header') {
      const headerMatch = body.match(/<text[^>]* name="([^"]*)"/)
      if (headerMatch) currentGroup = normalizeGroup(headerMatch[1])
      continue
    }

    const cells = [...body.matchAll(/<frame id="[^"]*" name="Cell"[^>]*>([\s\S]*?)<\/frame>/g)]
    if (cells.length < 2) continue

    const variableTexts = extractTextNames(cells[0][1])
    const variable = variableTexts[0]
    if (!variable || variable === 'Variable' || variable === 'Type') continue

    const typeTexts = extractTextNames(cells[1][1])
    const type = typeTexts[0] || 'unknown'

    const light = cells[2] ? parseValueCell(cells[2][1]) : { hex: null, step: null, raw: null }
    const dark = cells[3] ? parseValueCell(cells[3][1]) : { hex: null, step: null, raw: null }

    const token = buildTokenName(currentGroup, variable)
    const group = token.includes('/') ? token.split('/').slice(0, 2).join('/') : currentGroup || 'root'

    rows.push({
      token,
      group,
      type,
      light: light.hex || light.raw,
      lightStep: light.step,
      dark: dark.hex || dark.raw,
      darkStep: dark.step,
    })
  }

  return { table: tableName, count: rows.length, rows }
}

function main() {
  const xml = readFileSync(defaultInput, 'utf8')

  const tableBlocks = [...xml.matchAll(/<frame id="[^"]*" name="Table"[^>]*>([\s\S]*?)<\/frame>\s*(?=<frame id="[^"]*" name="Table"|<\/frame>\s*<\/frame>\s*$)/g)]

  // Fallback: split on Table frames inside Collection
  let tables = tableBlocks.map((m, i) => m[1])
  if (tables.length === 0) {
    const collection = xml.match(/<frame id="73102:139464"[\s\S]*/)?.[0] || xml
    tables = [...collection.matchAll(/<frame id="[^"]*" name="Table" x="[^"]*" y="[^"]*" width="[^"]*" height="[^"]*">([\s\S]*?)(?=<frame id="[^"]*" name="Table"|<\/frame>\s*<\/frame>\s*$)/g)].map(
      (m) => m[1],
    )
  }

  const tableNames = ['semantic-colors', 'color-primitives', 'size', 'effects']
  const parsed = {}

  tables.forEach((tableXml, i) => {
    const name = tableNames[i] || `table-${i}`
    parsed[name] = parseTable(tableXml, name)
  })

  // Group summary for semantic colors
  const semantic = parsed['semantic-colors']
  const groupCounts = {}
  for (const row of semantic?.rows || []) {
    const top = row.group.replace(/^--/, '')
    groupCounts[top] = (groupCounts[top] || 0) + 1
  }

  mkdirSync(OUT_DIR, { recursive: true })

  writeFileSync(join(OUT_DIR, 'semantic-colors.json'), JSON.stringify(semantic, null, 2))
  writeFileSync(join(OUT_DIR, 'color-primitives.json'), JSON.stringify(parsed['color-primitives'], null, 2))
  writeFileSync(join(OUT_DIR, 'size.json'), JSON.stringify(parsed['size'], null, 2))
  writeFileSync(join(OUT_DIR, 'effects.json'), JSON.stringify(parsed['effects'], null, 2))
  writeFileSync(
    join(OUT_DIR, 'meta.json'),
    JSON.stringify(
      {
        source: 'VDS Figma Collection 73102:139464',
        exportedAt: new Date().toISOString().slice(0, 10),
        counts: Object.fromEntries(Object.entries(parsed).map(([k, v]) => [k, v.count])),
        semanticGroups: groupCounts,
      },
      null,
      2,
    ),
  )

  console.log('Parsed token counts:')
  for (const [k, v] of Object.entries(parsed)) {
    console.log(`  ${k}: ${v.count}`)
  }
  console.log(`  semantic groups: ${Object.keys(groupCounts).length}`)
  console.log(`Written to ${OUT_DIR}`)
}

main()

File: scripts/parse-figma-collection.py
#!/usr/bin/env python3
"""Parse VDS Figma Collection metadata XML into token JSON files."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

DEFAULT_INPUT = Path.home() / (
    ".cursor/projects/Users-ankush-rustagi-Verkada-Code/agent-tools/"
    "9f913e5d-7a56-4475-8e89-b212a2181620.txt"
)
OUT_DIR = Path(__file__).resolve().parent.parent / "src/tokens/generated"

TABLE_IDS = [
    ("73102:139626", "semantic-colors"),
    ("73102:149124", "color-primitives"),
    ("73102:151299", "size"),
    ("73102:151952", "effects"),
]


def extract_table_xml(text: str, table_id: str) -> str:
    marker = f'<frame id="{table_id}" name="Table"'
    start = text.find(marker)
    if start < 0:
        return ""
    # Walk to matching close: next sibling Table or end of Collection inner content
    next_tables = [text.find(f'<frame id="{tid}" name="Table"') for tid, _ in TABLE_IDS if tid != table_id]
    next_tables = [i for i in next_tables if i > start]
    end = min(next_tables) if next_tables else len(text)
    return text[start:end]


def cell_texts(cell: str) -> list[str]:
    return re.findall(r'<text[^>]* name="([^"]*)"', cell)


def parse_value_cell(cell: str) -> dict:
    texts = [t for t in cell_texts(cell) if t not in ("Rectangle", "Variable", "Type", "Light", "Value", "Dark Mode Testing")]
    hex_match = re.search(r"\(#([0-9A-Fa-f]{6,8})\)", cell)
    if hex_match:
        return {"hex": f"#{hex_match.group(1).upper()}", "step": next((t for t in texts if re.fullmatch(r"\d+|transparent|scrim", t)), None)}
    # Direct hex like #FFFFFF00
    direct = next((t for t in texts if re.fullmatch(r"#[0-9A-Fa-f]{6,8}", t)), None)
    if direct:
        return {"hex": direct.upper(), "step": None}
    return {"hex": None, "step": None, "raw": " ".join(texts) if texts else None}


def normalize_group(header: str) -> str:
    h = header.strip()
    if h.startswith("--"):
        h = h[2:]
    return re.sub(r"\s*/\s*", "/", h)


def build_token(group: str, variable: str) -> str:
    if variable.startswith("--"):
        return variable
    if group:
        return f"--{group}/{variable}"
    return f"--{variable}"


def parse_semantic_table(table_xml: str) -> list[dict]:
    rows: list[dict] = []
    current_group = ""

    parts = re.split(r'(?=\n    <frame id="[^"]*" name="(?:Header|Row)")', table_xml)

    for part in parts:
        if 'name="Header"' in part[:80]:
            m = re.search(r'<text[^>]* name="([^"]*)"', part)
            if m and m.group(1) not in ("color",):
                current_group = normalize_group(m.group(1))
            continue

        if 'name="Row"' not in part[:80]:
            continue

        cells = re.findall(r'<frame id="[^"]*" name="Cell"[^>]*>([\s\S]*?)</frame>', part)
        if len(cells) < 2:
            continue

        var = cell_texts(cells[0])[0] if cell_texts(cells[0]) else None
        if not var or var in ("Variable", "Type"):
            continue

        token_type = cell_texts(cells[1])[0] if len(cells) > 1 and cell_texts(cells[1]) else "unknown"
        light = parse_value_cell(cells[2]) if len(cells) > 2 else {}
        dark = parse_value_cell(cells[3]) if len(cells) > 3 else {}

        token = build_token(current_group, var)
        parts_tok = token.lstrip("-").split("/")
        group_key = "/".join(parts_tok[:2]) if len(parts_tok) >= 2 else (parts_tok[0] if parts_tok else "root")

        rows.append(
            {
                "token": token,
                "group": group_key,
                "type": token_type,
                "light": light.get("hex") or light.get("raw"),
                "lightStep": light.get("step"),
                "dark": dark.get("hex") or dark.get("raw"),
                "darkStep": dark.get("step"),
            }
        )

    return rows


def parse_simple_table(table_xml: str, mode: str) -> list[dict]:
    rows: list[dict] = []
    current_palette = ""

    parts = re.split(r'(?=\n    <frame id="[^"]*" name="(?:Header|Row)")', table_xml)

    for part in parts:
        if 'name="Header"' in part[:80]:
            m = re.search(r'<text[^>]* name="([^"]*)"', part)
            if m:
                name = m.group(1)
                if name not in ("--color primitives", "size", "effects", "icon", "radius", "space", "stroke", "blur", "opacity"):
                    current_palette = normalize_group(name)
                elif name.startswith("--"):
                    current_palette = normalize_group(name)
            continue

        if 'name="Row"' not in part[:80]:
            continue

        cells = re.findall(r'<frame id="[^"]*" name="Cell"[^>]*>([\s\S]*?)</frame>', part)
        if len(cells) < 2:
            continue

        var = cell_texts(cells[0])[0] if cell_texts(cells[0]) else None
        if not var or var in ("Variable", "Type", "Value"):
            continue

        token_type = cell_texts(cells[1])[0] if cell_texts(cells[1]) else mode
        value = parse_value_cell(cells[2]) if len(cells) > 2 else parse_value_cell(cells[1])

        if mode == "primitives":
            token = f"--color-primitives/{current_palette}/{var}" if current_palette else f"--color-primitives/{var}"
        else:
            token = f"--{current_palette}/{var}" if current_palette else f"--{var}"

        rows.append(
            {
                "token": token,
                "palette": current_palette or "root",
                "type": token_type,
                "value": value.get("hex") or value.get("raw"),
                "step": value.get("step"),
            }
        )

    return rows


def main() -> None:
    input_path = Path(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT_INPUT
    text = input_path.read_text()
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    semantic_xml = extract_table_xml(text, "73102:139626")
    primitives_xml = extract_table_xml(text, "73102:149124")
    size_xml = extract_table_xml(text, "73102:151299")
    effects_xml = extract_table_xml(text, "73102:151952")

    semantic = parse_semantic_table(semantic_xml)
    primitives = parse_simple_table(primitives_xml, "primitives")
    size = parse_simple_table(size_xml, "size")
    effects = parse_simple_table(effects_xml, "effects")

    group_counts: dict[str, int] = {}
    for row in semantic:
        group_counts[row["group"]] = group_counts.get(row["group"], 0) + 1

    meta = {
        "source": "VDS Figma Collection 73102:139464",
        "exportedAt": "2026-05-22",
        "counts": {
            "semantic-colors": len(semantic),
            "color-primitives": len(primitives),
            "size": len(size),
            "effects": len(effects),
        },
        "semanticGroups": dict(sorted(group_counts.items(), key=lambda x: (-x[1], x[0]))),
    }

    (OUT_DIR / "semantic-colors.json").write_text(json.dumps({"rows": semantic, **meta}, indent=2))
    (OUT_DIR / "color-primitives.json").write_text(json.dumps({"rows": primitives, **meta}, indent=2))
    (OUT_DIR / "size.json").write_text(json.dumps({"rows": size, **meta}, indent=2))
    (OUT_DIR / "effects.json").write_text(json.dumps({"rows": effects, **meta}, indent=2))
    (OUT_DIR / "meta.json").write_text(json.dumps(meta, indent=2))

    print("Exported:")
    for k, v in meta["counts"].items():
        print(f"  {k}: {v}")
    print(f"  groups: {len(group_counts)}")


if __name__ == "__main__":
    main()
